src/metabase/api/dashboard.clj

```diff
@@ -598,13 +600,90 @@
                             :num-tabs       (count created-tab-ids)
                             :total-num-tabs total-num-tabs})))
 
+;;;;;;;;;;;; Bad pulse check & repair
+
+(defn- bad-pulse-notification-data
+  "Given a pulse and bad parameters, return relevant notification data:
+  - The name of the pulse
+  - Which selected parameter values are broken
+  - The user info for the creator of the pulse
+  - The users affected by the pulse"
+  [{bad-pulse-id :id pulse-name :name :keys [parameters creator_id]}]
+  (let [creator                  (t2/select-one [:model/User :first_name :last_name :email] creator_id)
+        pulse-channel-ids        (mapv :id (t2/select [:model/PulseChannel :id] :pulse_id [:= bad-pulse-id]))
+        ;; When checks are defensive as in () causes SQL exceptions
+        pulse-channel-recipients (when (seq pulse-channel-ids)
+                                   (t2/select :model/PulseChannelRecipient :pulse_channel_id [:in pulse-channel-ids]))
+        affected-users           (when (seq pulse-channel-recipients)
+                                   (t2/select [:model/User :first_name :last_name :email]
+                                     :id [:in (map :user_id pulse-channel-recipients)]))]
```

Comment:
can we send pulses to emails that are not associated with a user? I ask because if we can do that, I then wonder if it's possible that `(map :user_id pulse-channel-recipients)` is an empty seq. If so, the defensive `when` might still break in such a case?

Reply:
At this point, we just send the user name or slack channel if they are affected, so I think we're good to go. If the seq returns nil, the email will just not contain a list of users affected, but the owners will still be notified.